forge-game/src/main/java/forge/game/staticability/StaticAbilityPanharmonicon.java

```diff
@@ -156,10 +159,55 @@ public static boolean applyPanharmoniconAbility(final StaticAbility stAb, final
             if (!stAb.matchesValidParam("ValidActivator", sa.getActivatingPlayer())) {
                 return false;
             }
-        } else if (trigMode.equals(TriggerType.DamageDone) || trigMode.equals(TriggerType.DamageDoneOnce)) {
-            if (!stAb.matchesValidParam("ValidTarget", runParams.get(AbilityKey.DamageTarget))) {
+        } else if (trigMode.equals(TriggerType.DamageDone) || trigMode.equals(TriggerType.DamageDoneOnce) 
+                || trigMode.equals(TriggerType.DamageAll) || trigMode.equals(TriggerType.DamageDealtOnce)) {
+            if (stAb.hasParam("CombatDamage") && stAb.getParam("CombatDamage").equalsIgnoreCase("True") != 
+                    (Boolean) runParams.get(AbilityKey.IsCombatDamage)) {
                 return false;
             }
+            if (trigMode.equals(TriggerType.DamageDone)) {
+                if (!stAb.matchesValidParam("ValidSource", runParams.get(AbilityKey.DamageSource))) {
+                    return false;
+                }
+                if (!stAb.matchesValidParam("ValidTarget", runParams.get(AbilityKey.DamageTarget))) {
+                    return false;
+                }
+            }
+            if (trigMode.equals(TriggerType.DamageDoneOnce)) {
+                if (!stAb.matchesValidParam("ValidTarget", runParams.get(AbilityKey.DamageTarget))) {
+                    return false;
+                }
+                Map<Card, Integer> dmgMap = (Map<Card, Integer>) runParams.get(AbilityKey.DamageMap);
+                // 1. check it's valid cause for static
+                // 2. and it must also be valid for trigger event
+                if (!Iterables.any(dmgMap.keySet(), Predicates.and(
+                        CardPredicates.matchesValidParam(stAb, "ValidSource"),
+                        CardPredicates.matchesValidParam(trigger, "ValidSource")
+                        ))) {
+                    return false;
+                }
+                // DamageAmount$ can be ignored for now (its usage doesn't interact with ValidSource from either)
+            }
+            if (trigMode.equals(TriggerType.DamageDealtOnce)) {
+                if (!stAb.matchesValidParam("ValidSource", runParams.get(AbilityKey.DamageSource))) {
+                    return false;
+                }
+                Map<Card, Integer> dmgMap = (Map<Card, Integer>) runParams.get(AbilityKey.DamageMap);
+                if (!Iterables.any(dmgMap.keySet(), Predicates.and(
+                        CardPredicates.matchesValidParam(stAb, "ValidTarget"),
+                        CardPredicates.matchesValidParam(trigger, "ValidTarget")
+                        ))) {
+                    return false;
+                }
```

Comment:
@tool4ever DamageDealtOnce uses this Map `Map<GameEntity, Integer>`

meaning that `CardPredicates.matchesValidParam` should probably be moved to `GameObjectPredicates.matchesValidParam`

Reply:
right, good catch